src/main/java/com/teragrep/cfe_39/consumers/kafka/ReadCoordinator.java

```diff
@@ -0,0 +1,105 @@
+        this.readerKafkaProperties = readerKafkaProperties;
+        this.callbackFunction = callbackFunction;
+        this.hdfsStartOffsets = hdfsStartOffsets;
+    }
+
+    private KafkaReader createKafkaReader(Properties readerKafkaProperties,
+                                          String topic,
+                                          Consumer<List<RecordOffset>> callbackFunction,
+                                          boolean useMockKafkaConsumer) {
+
+        org.apache.kafka.clients.consumer.Consumer<byte[], byte[]> kafkaConsumer;
+        if (useMockKafkaConsumer) { // Test mode is on, create mock consumers with assigned partitions that are not overlapping with each other.
+            String name = Thread.currentThread().getName(); // Use thread name to identify which thread is running the code.
+            if (Objects.equals(name, "testConsumerTopic1")) {
+                kafkaConsumer = MockKafkaConsumerFactoryTemp.getConsumer(1); // creates a Kafka MockConsumer that has the odd numbered partitions assigned to it.
+            }else {
+                kafkaConsumer = MockKafkaConsumerFactoryTemp.getConsumer(2); // creates a Kafka MockConsumer that has the even numbered partitions assigned to it.
+            }
+        } else { // Test mode is off, subscribe method should handle assigning the partitions automatically to the consumer based on group id parameters of readerKafkaProperties.
+            kafkaConsumer = new KafkaConsumer<>(readerKafkaProperties, new ByteArrayDeserializer(), new ByteArrayDeserializer());
+            kafkaConsumer.subscribe(Collections.singletonList(topic));
+        }
+
+        Set<TopicPartition> assignment = kafkaConsumer.assignment();
+        // Seek the consumer to topic partition offset defined by the latest record that is committed to HDFS.
+        for (TopicPartition topicPartition : assignment) {
+            if (hdfsStartOffsets.containsKey(topicPartition)) {
+                long position = kafkaConsumer.position(topicPartition);
+                if (position < hdfsStartOffsets.get(topicPartition)) {
+                    kafkaConsumer.seek(topicPartition, hdfsStartOffsets.get(topicPartition));
+                }
+            }
+        }
+
+        return new KafkaReader(kafkaConsumer, callbackFunction);
+    }
+
+    // Part or Runnable implementation, called when the thread is started.
+    @Override
+    public void run() {
+        boolean useMockKafkaConsumer = Boolean.parseBoolean(readerKafkaProperties.getProperty("useMockKafkaConsumer", "false"));
+        try (
+                KafkaReader kafkaReader = createKafkaReader(
+                        readerKafkaProperties,
+                        queueTopic,
+                        callbackFunction,
+                        useMockKafkaConsumer
+                )
+        ) {
+            while (run) {
+                kafkaReader.read();
+            }
+        }
+    }
+
+    // remove?
```

Comment:
Add FIXME so it can be automatically discovered

Reply:
Solved in commit 0d536bee72cbc9b97b70dd731f794381fadc3e1f